daemon/internal/cloud_client/cloud_client.go

```diff
@@ -0,0 +1,199 @@
+		for event := range client.eventChannel {
+			event.handle(client)
+		}
+	}()
+
+	// Start the heartbeat connection to the cloud.
+	client.eventChannel <- newHeartbeatEvent()
+
+	return client.eventChannel
+}
+
+func (client *cloudClient) newLearnClient(serviceID akid.ServiceID) rest.LearnClient {
+	return rest.NewLearnClient(client.host, client.clientID, serviceID)
+}
+
+func (client *cloudClient) getInfo(serviceID akid.ServiceID, traceID akid.LearnSessionID) (*serviceInfo, *traceInfo) {
+	serviceInfo, ok := client.serviceInfoByID[serviceID]
+	if !ok {
+		return nil, nil
+	}
+
+	traceInfo, ok := serviceInfo.traces[traceID]
+	if !ok {
+		return serviceInfo, nil
+	}
+
+	return serviceInfo, traceInfo
+}
+
+// Registers the service with the daemon if needed. Upon registration, a
+// longPollServiceEvent is scheduled for the service.
+func (client *cloudClient) ensureServiceRegistered(serviceID akid.ServiceID) *serviceInfo {
+	if serviceInfo, ok := client.serviceInfoByID[serviceID]; ok {
+		// Service already registered.
+		return serviceInfo
+	}
+
+	// Register the new service and schedule a longPollServiceEvent.
+	serviceInfo := client.newServiceInfo(serviceID)
+	client.serviceInfoByID[serviceID] = serviceInfo
+	client.eventChannel <- newLongPollServiceEvent(serviceID)
+
+	return serviceInfo
+}
+
+// Determines which traces are being collected for the given service.
+func (client *cloudClient) getCurrentTraces(serviceID akid.ServiceID) []akid.LearnSessionID {
+	result := []akid.LearnSessionID{}
+	for traceID := range client.serviceInfoByID[serviceID].traces {
+		result = append(result, traceID)
+	}
+	return result
+}
+
+// Starts a goroutine for collecting trace events and sending them to the
+// cloud. Assumes the given service ID has been registered.
+func (client *cloudClient) startTraceEventCollector(serviceID akid.ServiceID, loggingOptions daemon.LoggingOptions) {
+	serviceInfo, traceInfo := client.getInfo(serviceID, loggingOptions.TraceID)
+	if serviceInfo == nil {
+		printer.Debugf("Got a new trace from the cloud for an unregistered service: %q\n", akid.String(serviceID))
+		return
+	}
+
+	if traceInfo != nil {
+		if traceInfo.active {
+			printer.Debugf("Got an allegedly new trace from the cloud, but already collecting events for that trace: %q\n", akid.String(loggingOptions.TraceID))
+		}
+
+		// Reactivate the trace and update its logging options.
+		traceInfo.active = true
+		traceInfo.loggingOptions = loggingOptions
+		return
+	}
+
+	// Start a collector goroutine.
+	learnClient := serviceInfo.learnClient
+	filterThirdPartyTrackers := loggingOptions.FilterThirdPartyTrackers
+	traceEventChannel := make(chan *TraceEvent, TRACE_BUFFER_SIZE)
+	go func() {
+		// Create the collector.
+		collector := trace.NewBackendCollector(serviceID, loggingOptions.TraceID, learnClient, api_schema.Inbound, client.plugins, nil)
+		if filterThirdPartyTrackers {
+			collector = trace.New3PTrackerFilterCollector(collector)
+		}
+		defer collector.Close()
```

Comment:
This goroutine defined inline is too big, can it be feasibly made into its own function?

Reply:
Feasible, sure. But where would it go? I'd like to avoid having an externally function visible that shouldn't be called anywhere else.

Reply:
The place I'm coming from is that inevitably somebody wants to write a unit test and the inline function is too hard to get to with the right arguments or mocking.  (I've been that person, with this example: https://github.com/hashicorp/vault/blob/2a8c36be280da6d0a1303c814a8d05ce274e5414/vault/acl.go#L523)

It can be package-scoped (not exported) so it's not entirely "visible".

Breaking it out also makes it easier to interpret backtraces, particularly if there are multiple funcs() involved.

Reply:
Done in 9eca8de.